src/crypto/PSAOperationalKeystore.cpp

```diff
@@ -209,5 +273,31 @@ void PSAOperationalKeystore::ReleasePendingKeypair()
     mIsPendingKeypairActive = false;
 }
 
+CHIP_ERROR PSAOperationalKeystore::MigrateOpKeypairForFabric(FabricIndex fabricIndex,
+                                                             OperationalKeystore & operationalKeystore) const
+{
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndex), CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+    P256SerializedKeypair serializedKeypair;
+
+    // Do not allow overwriting the existing key and just remove it from the previous Operational Keystore if needed.
+    if (!HasOpKeypairForFabric(fabricIndex))
+    {
+        ReturnErrorOnFailure(operationalKeystore.ExportOpKeypairForFabric(fabricIndex, serializedKeypair));
+
+        PersistentP256Keypair keypair(fabricIndex);
+        ReturnErrorOnFailure(keypair.Deserialize(serializedKeypair));
+
+        // Migrated key is not useful anymore, remove it from the previous keystore.
+        ReturnErrorOnFailure(operationalKeystore.RemoveOpKeypairForFabric(fabricIndex));
+    }
+    else if (operationalKeystore.HasOpKeypairForFabric(fabricIndex))
+    {
+        ReturnErrorOnFailure(operationalKeystore.RemoveOpKeypairForFabric(fabricIndex));
+    }
```

Comment:
What does the else cover? Is it tested? This just looks like it deletes a key, but only the other block above did a migration.

Reply:
If the new Operational Keystore already contains the key for a given fabric, the attempt to replace it should be forbidden, and we should remove the old one. It prevents from attack that someone can write directly the key to the obsolete keystore and in this way replace the converted one. And regarding tests, yes, it is tested in https://github.com/project-chip/connectedhomeip/pull/31279/files#diff-2e13b279258becb5810bff25892cd615de76dac46ba1ba1fbb78fcbd4450edebR369 and https://github.com/project-chip/connectedhomeip/pull/31279/files#diff-2cab20d625cc0e0133ed42f0de6b8a30e40dfd529490f80d7730e4f8cff42839R252 .